File: alchemist-incarnation-sapere/src/main/java/it/unibo/alchemist/model/implementations/timedistributions/SAPEREExponentialTime.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.implementations.timedistributions;

import it.unibo.alchemist.model.sapere.dsl.impl.Expression;
import it.unibo.alchemist.model.sapere.dsl.IExpression;
import it.unibo.alchemist.model.sapere.dsl.ITreeNode;
import it.unibo.alchemist.model.timedistributions.ExponentialTime;
import org.apache.commons.math3.random.RandomGenerator;

import it.unibo.alchemist.model.times.DoubleTime;
import it.unibo.alchemist.model.interfaces.ILsaMolecule;
import it.unibo.alchemist.model.Time;
import org.danilopianini.lang.HashString;

import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.Semaphore;

/**
 * Allows for a Markovian event whose lambda is computed dynamically using a
 * rate equation.
 * 
 */
public final class SAPEREExponentialTime extends ExponentialTime<List<ILsaMolecule>> implements SAPERETimeDistribution {

    /**
     * 
     */
    private static final long serialVersionUID = -687039899173488373L;
    private static final String F_PATTERN = "###.######################";
    private static final DecimalFormat FORMAT = new DecimalFormat(F_PATTERN, DecimalFormatSymbols.getInstance(Locale.ENGLISH));
    private static final Semaphore FORMAT_MUTEX = new Semaphore(1);

    private final IExpression exp;
    private final double staticRate;
    private final boolean numericRate;
    private Map<HashString, ITreeNode<?>> matches;

    /**
     * @param rateEquation
     *            the rate equation
     * @param random
     *            the {@link RandomGenerator}
     */
    public SAPEREExponentialTime(final String rateEquation, final RandomGenerator random) {
        this(rateEquation, new DoubleTime(), random);
    }

    /**
     * @param rateEquation
     *            the rate equation
     * @param start
     *            initial time
     * @param random
     *            the {@link RandomGenerator}
     */
    public SAPEREExponentialTime(final String rateEquation, final Time start, final RandomGenerator random) {
        super(Double.NaN, start, Objects.requireNonNull(random));
        double temp = 0d;
        boolean numeric = true;
        try {
            temp = Double.parseDouble(Objects.requireNonNull(rateEquation));
        } catch (NumberFormatException e) {
            numeric = false;
        }
        numericRate = numeric;
        staticRate = temp;
        if (numericRate) {
            if (Double.isInfinite(staticRate)) {
                exp = new Expression("asap");
            } else {
                try {
                    FORMAT_MUTEX.acquireUninterruptibly();
                    exp = new Expression(FORMAT.format(staticRate)); // NOPMD: access is synchronized via mutex
                } finally {
                    FORMAT_MUTEX.release();
                }
            }
        } else {
            exp = new Expression(rateEquation);
        }
    }

    @Override
    public double getRate() {
        return numericRate ? staticRate : (Double) exp.calculate(matches).getValue(matches);
    }

    @Override
    public boolean isStatic() {
        return numericRate;
    }

    @Override
    public void setMatches(final Map<HashString, ITreeNode<?>> match) {
        if (!numericRate) {
            matches = match;
        }
    }

    @Override
    public IExpression getRateEquation() {
        return exp;
    }

}

File: alchemist-incarnation-sapere/src/main/kotlin/it/unibo/alchemist/model/sapere/timedistributions/SAPEREExponentialTime.kt
/*
 * Copyright (C) 2010-2026, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.model.sapere.timedistributions

import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.Time
import it.unibo.alchemist.model.TimeDistribution
import it.unibo.alchemist.model.sapere.ILsaMolecule
import it.unibo.alchemist.model.sapere.dsl.IExpression
import it.unibo.alchemist.model.sapere.dsl.ITreeNode
import it.unibo.alchemist.model.sapere.dsl.impl.Expression
import it.unibo.alchemist.model.timedistributions.ExponentialTime
import it.unibo.alchemist.model.times.DoubleTime
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale
import java.util.concurrent.Semaphore
import org.apache.commons.math3.random.RandomGenerator
import org.danilopianini.lang.HashString

/** An exponential generator whose rate is computed from a SAPERE match expression. */
@JvmSuppressWildcards
class SAPEREExponentialTime(private val rateEquation: String, start: Time, randomGenerator: RandomGenerator) :
    ExponentialTime<@JvmSuppressWildcards List<ILsaMolecule>>(Double.NaN, start, randomGenerator),
    SAPERETimeDistribution {

    /** Builds a generator starting at time zero. */
    constructor(rateEquation: String, randomGenerator: RandomGenerator) :
        this(rateEquation, DoubleTime(), randomGenerator)

    private val staticRate = rateEquation.toDoubleOrNull() ?: 0.0
    private val numericRate = rateEquation.toDoubleOrNull() != null
    private val expression: IExpression = if (numericRate) {
        if (staticRate.isInfinite()) {
            Expression("asap")
        } else {
            FORMAT_MUTEX.acquireUninterruptibly()
            try {
                Expression(FORMAT.format(staticRate))
            } finally {
                FORMAT_MUTEX.release()
            }
        }
    } else {
        Expression(rateEquation)
    }
    private var matches: Map<HashString, ITreeNode<*>>? = null

    override fun getRate(): Double = if (numericRate) {
        staticRate
    } else {
        expression.calculate(matches).getValue(matches) as Double
    }

    override val lambda: Double get() = getRate()

    override fun sample(): Time = genTime(getRate())

    override fun newInstanceOn(node: Node<List<ILsaMolecule>>): TimeDistribution<List<ILsaMolecule>> =
        SAPEREExponentialTime(rateEquation, startTime, randomGenerator)

    override fun isStatic(): Boolean = numericRate

    override fun setMatches(match: Map<HashString, ITreeNode<*>>) {
        if (!numericRate) {
            matches = match
        }
    }

    override fun getRateEquation(): IExpression = expression

    private companion object {
        private const val FORMAT_PATTERN = "###.######################"
        private val FORMAT = DecimalFormat(FORMAT_PATTERN, DecimalFormatSymbols.getInstance(Locale.ENGLISH))
        private val FORMAT_MUTEX = Semaphore(1)
    }
}
